### PR TITLE
Tommy/unit tests player

### PR DESCRIPTION
**Pull request by teelau**

### CODE REVIEW

File: server/models/player_test.go [hunk 1]
```diff
@@ -1,6 +1,10 @@
 package models
 
-import "testing"
+import (
+	"testing"
+
+	"github.com/gorilla/websocket"
+)
 
 func TestAddPoints(t *testing.T) {
```
Comment (brian-nguyen) [inline server/models/player_test.go]: This test can be expanded

File: server/models/player_test.go [hunk 2]
```diff
@@ -10,3 +14,114 @@ func TestAddPoints(t *testing.T) {
...
+	testHeight1 := 10.0
+	testWidth1 := 20.0
+	testAngle1 := 0.0
+	testPosition1 := Position{5, 5}
+	p := new(Player)
+	p.Position = testPosition1
+	p.Angle = testAngle1
+	p.Controls.Left = false
+	p.Controls.Right = false
+	p.Controls.Up = false
```
[the remark this answers is not in the document]
Comment (brian-nguyen) [reply server/models/player_test.go]: When allocating a struct using new, the values in the struct are zero-initialized. For booleans, this value is false. So, you can save yourself some lines by not having to explicitly set `p.Controls.Left = false`

File: server/models/player_test.go [hunk 3]
```diff
@@ -10,3 +14,114 @@ func TestAddPoints(t *testing.T) {
 		t.Error("Error adding points")
 	}
 }
+
+func TestCreatePlayer(t *testing.T) {
```
Comment (brian-nguyen) [inline server/models/player_test.go]: This test can be omitted if you make use of the CreatePlayer function in your tests. You can supply nil to the WebSocket param if you want to ignore anything related to sockets

File: server/models/player_test.go [hunk 4]
```diff
@@ -10,3 +14,114 @@ func TestAddPoints(t *testing.T) {
...
+	}
+	p.Controls.Right = false
+	p.Controls.Up = true
+	p.UpdatePosition(testHeight1, testWidth1)
+	if p.Angle != (testAngle1) {
+		t.Error("Error in up key player control")
+	}
+	p.UpdatePosition(testHeight1, testWidth1)
+
+}
```
Comment (brian-nguyen) [inline server/models/player_test.go]: This test only considers a single key press in the left and right direction. Although the coverage is great, it doesn't consider enough meaningful behaviour. We should try and write tests for actual behaviour that we want to see when playtesting. How about testing if a Player can rotate 360 degrees? A test case like that would tell us that the Player can move in any direction on the arena. How about UpdatePosition near boundaries?

File: server/models/player_test.go [hunk 5]
```diff
@@ -10,3 +14,114 @@ func TestAddPoints(t *testing.T) {
...
+	p.hitJunk()
+	if p.Velocity.Dx != testVelocity.Dx*JunkBounceFactor {
+		t.Error("Error calculating player Dx hitting junk")
+	}
+	if p.Velocity.Dy != testVelocity.Dy*JunkBounceFactor {
+		t.Error("Error calculating player Dy hitting junk")
+	}
+}
+
+func TestKeyDownHandler(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (brian-nguyen) [reply server/models/player_test.go]: Also camel case the function name

[the remark this answers is not in the document]
Comment (teelau) [reply server/models/player_test.go]: camel case is not good for tests, they prevent the testing tool tips from appearing

File: server/models/player_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (teelau) [reply server/models/player_test.go]: i think values should be local to the tests so they are as isolated as possible unless we start making a test api

File: server/models/player_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (brian-nguyen) [reply server/models/player_test.go]: Oops by camel case, I mean Upper CamelCase aka Pascal Case aka MixedCaps according to the GoDoc

File: server/models/player_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (brian-nguyen) [reply server/models/player_test.go]: See @Grafne23 and @SeeratSek's tests. They do a good job of figuring out what values to share across all tests and what to keep local. In Graeme's case, unless he was specifically testing a position/velocity, he chose to have all his tests run with a default position/velocity in the center, and even those test velocities that differ are actually variations of default value. In both cases, Graeme and Seerat also defined multiple sets of test values for a particular test locally. It all depends on the situation, and I think there is room for improvement in this test.

**Pull request closed without merging**